### PR TITLE
[panopto] Extract slides + subtitles for videos without m3u8 subs

### PR DESCRIPTION
**Pull request by coletdjnz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Panopto provides captions in json via an endpoint. 
Note the duration of the subtitle is not always available, so I use the next captions start time as the end in that case.

Also added support for downloading slides if available, plus chapter previews.

### CODE REVIEW

**Comment by coletdjnz** [comment]
Tests I've added current do not entirely pass. We need to add functionality to `InfoExtractor._merge_subtitles` to support merging subtitles that have `data` instead of `url`.

**Comment by coletdjnz** [comment]
needs https://github.com/yt-dlp/yt-dlp/pull/3044

File: test/test_subtitles.py [hunk 1]
```diff
@@ -410,5 +411,39 @@ def test_subtitles_sami_format(self):
...
+        self.assertEqual(md5(subtitles['en']), 'ade23c2bad3dcae5ce4e27527922b4e9')
+
+    def test_no_subtitles(self):
+        self.url = 'https://demo.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=495ac261-8cf8-4ea5-9714-ad4701326a1b'
+        self.DL.params['writesubtitles'] = True
+        self.DL.params['allsubtitles'] = True
+        subtitles = self.getSubtitles()
+        self.assertFalse(subtitles)
+
+
```
Comment (pukkandan) [inline test/test_subtitles.py:447]: Can't a extractor download test check these? I'd prefer not to have extractors define their own additional tests. These often tend to get left behind when extractor is updated

Comment (coletdjnz) [reply test/test_subtitles.py:447]: I'm not sure what the standard way is. Wouldn't I have to duplicate a test a few times?

File: yt_dlp/extractor/panopto.py [hunk 2]
```diff
@@ -178,19 +253,82 @@ def _mark_watched(self, base_url, video_id, delivery_info):
...
+            duration = timestamp.get('Duration')
+            obj_id, obj_sn = timestamp.get('ObjectIdentifier'), timestamp.get('ObjectSequenceNumber'),
+            if timestamp.get('EventTargetType') == 'PowerPoint' and obj_id is not None and obj_sn is not None:
+                image_frags.setdefault('slides', []).append({
+                    'url': base_url + f'/Pages/Viewer/Image.aspx?id={obj_id}&number={obj_sn}',
+                    'duration': duration
+                })
+
+            obj_pid, session_id, abs_time = timestamp.get('ObjectPublicIdentifier'), timestamp.get('SessionID'), timestamp.get('AbsoluteTime')
+            if all(v is not None for v in (obj_pid, session_id, abs_time)):
```
Comment (pukkandan) [inline yt_dlp/extractor/panopto.py:283]:
```suggestion
            if None not in (obj_pid, session_id, abs_time):
```

**Pull request merged**